### PR TITLE
[WIP] Subgraph

### PR DESCRIPTION
**Pull request by mbrookhart**

### CODE REVIEW

File: 3rdparty/ngraph_bridge/src/ngraph_utils.h [hunk 1]
```diff
@@ -55,9 +55,8 @@ inline bool ngraph_log_viz() {
 inline bool ngraph_log_timer() {
   return dmlc::GetEnv("MXNET_NGRAPH_TIMER", false);
 }
-inline bool ngraph_log_verbose_detail() {
-  return dmlc::GetEnv("MXNET_NGRAPH_VERBOSE_DETAIL", false);
-}
+const bool ngraph_log_verbose_detail =
+    dmlc::GetEnv("MXNET_NGRAPH_VERBOSE_DETAIL", false);
```
[the remark this answers is not in the document]
Comment (ashokei) [reply 3rdparty/ngraph_bridge/src/ngraph_utils.h]: this change would make it read it at libmxnet.so loadtime, and reuse; function would re-read everytime.

[the remark this answers is not in the document]
Comment (mbrookhart) [reply 3rdparty/ngraph_bridge/src/ngraph_utils.h]: For this env variable, I guess that's okay, others we want to read each time I think

File: 3rdparty/ngraph_bridge/src/ngraph_graph.h [hunk 2]
```diff
@@ -187,7 +187,7 @@ inline std::shared_ptr<ngraph::runtime::Backend> GetBackendFromContext(
 
 class OutputElement;
 
-using MapEntry = std::pair<nnvmNodePtr, size_t>;
+using MapEntry = std::pair<const nnvm::Node *, size_t>;
```
[the remark this answers is not in the document]
Comment (ashokei) [reply 3rdparty/ngraph_bridge/src/ngraph_graph.h]: the key of Map was shared_ptr, as opposed to real pointer. This change allows to map from real address of node, as opposed to a share_ptr of a node (which we may not have, but we have the real address of node)

File: 3rdparty/ngraph_bridge/src/ngraph_graph.h [hunk 3]
```diff
@@ -235,13 +235,13 @@ class Graph : public Node {
 
   // Add a node to the graph
   void AddNode(NodePtr node) {
-    entry_map_[MapEntry{node->orig_node_, node->multi_output_index_}] = node;
+    entry_map_[MapEntry{node->orig_node_.get(), node->multi_output_index_}] = node;
```
[the remark this answers is not in the document]
Comment (ashokei) [reply 3rdparty/ngraph_bridge/src/ngraph_graph.h]: this is only used as key (and const) , not value. We do not expect this to be used other than as a key.

File: 3rdparty/ngraph_bridge/src/ngraph_subgraph.cc [hunk 4]
```diff
@@ -44,12 +41,19 @@ class NgraphSubgraphOperator {
...
+  nnvm::NodeAttrs attrs_;
 };
 
 void NgraphSubgraphOperator::Forward(const OpContext& ctx,
                                      const std::vector<NDArray>& inputs,
                                      const std::vector<OpReqType>& req,
                                      const std::vector<NDArray>& outputs) {
+  // check and init ngraph_
+  if (!ngraph_) {
+      NGImperative ngi(*attrs_.subgraphs[0], ctx.run_ctx.ctx, inputs, &req, outputs);
```
[the remark this answers is not in the document]
Comment (ashokei) [reply 3rdparty/ngraph_bridge/src/ngraph_subgraph.cc]: they seem to create separate Subgraph Node for each subgraph, so currently expecting only one subgraph here (we can double check)

File: 3rdparty/ngraph_bridge/src/ngraph_subgraph.h [hunk 5]
```diff
@@ -27,39 +28,54 @@ using namespace mxnet::op;
...
 
   bool SelectOutput(const nnvm::Node &n, const nnvm::Node &new_node) override {
+    return (!n.is_variable() && is_node_selected(new_node));
+  }
+
+ private:
+  Compiler compiler_;
+  bool is_node_selected(const nnvm::Node &n) {
+    NodePtr nn;
+    MapEntry tmp{&n, 0};
```
Comment (mbrookhart) [inline 3rdparty/ngraph_bridge/src/ngraph_subgraph.h]: Oh, I see, the subgraph API is forcing const ref, you don't have a shared pointer.

**Comment by mbrookhart** [comment]
Closed in favor of #403

**Pull request closed without merging**